File: apps/samples/src/image_predictor.cpp
#include "image_predictor.hpp"

void ImagePredictor::predict(int rows, int cols, unsigned char *buffer){
    cout<<"image processing start"<<endl;
    Mat frame ( rows, cols, CV_8UC3, buffer);
    cout<< "frame dims : ("<< frame.rows<<","<< frame.cols <<")"<<endl;

    //cout<< frame <<endl;

    vector<vector<int>> bboxes;
    Mat frameFace;

    Mat frameOpenCVDNN = frame.clone();
    int frameHeight = frameOpenCVDNN.rows;
    int frameWidth = frameOpenCVDNN.cols;

    double inScaleFactor = 1.0;
    Size size = Size(300, 300);
    Scalar meanVal = Scalar(104, 117, 123);

    Mat inputBlob;
    inputBlob = cv::dnn::blobFromImage(frameOpenCVDNN, inScaleFactor, size, meanVal, true, false);
    cout<<"input Blob : " << inputBlob.dims <<endl;
    faceNet.setInput(inputBlob, "data");
    Mat detection = faceNet.forward("detection_out");
    Mat detectionMat(detection.size[2], detection.size[3], CV_32F, detection.ptr<float>());
    for(int i = 0; i < detectionMat.rows; i++)
    {
        float confidence = detectionMat.at<float>(i, 2);
        if(confidence > 0.5)
        {
            int x1 = static_cast<int>(detectionMat.at<float>(i, 3) * frameWidth);
            int y1 = static_cast<int>(detectionMat.at<float>(i, 4) * frameHeight);
            int x2 = static_cast<int>(detectionMat.at<float>(i, 5) * frameWidth);
            int y2 = static_cast<int>(detectionMat.at<float>(i, 6) * frameHeight);
            vector<int> box = {x1, y1, x2, y2};
            bboxes.push_back(box);
            rectangle(frameOpenCVDNN, Point(x1, y1), Point(x2, y2), Scalar(0, 255, 0),2, 4);
        }
    }
    cout<<"done? "<<endl;
    tie(frameFace, bboxes) = make_tuple(frameOpenCVDNN, bboxes);
    if(bboxes.size() == 0) {
        cout << "No face detected, checking next frame." << endl;
    } else {
        cout << "Face detected!" << endl;
        for (auto it = begin(bboxes); it != end(bboxes); ++it) {
            Rect rec(it->at(0) - padding, it->at(1) - padding, it->at(2) - it->at(0) + 2*padding, it->at(3) - it->at(1) + 2*padding);
            Mat face = frame(rec); // take the ROI of box on the frame

            Mat blob;
            blob = blobFromImage(face, 1, Size(227, 227), MODEL_MEAN_VALUES, false);
            genderNet.setInput(blob);
            vector<float> genderPreds = genderNet.forward();
            int max_index_gender = std::distance(genderPreds.begin(), max_element(genderPreds.begin(), genderPreds.end()));
            string gender = genderList[max_index_gender];
            cout << "Gender: " << gender << endl;
            ageNet.setInput(blob);
            vector<float> agePreds = ageNet.forward();
            int max_indice_age = std::distance(agePreds.begin(), max_element(agePreds.begin(), agePreds.end()));
            string age = ageList[max_indice_age];
            cout << "Age: " << age << endl;
            string label = gender + ", " + age; // label
            cv::putText(frameFace, label, Point(it->at(0), it->at(1) -15), cv::FONT_HERSHEY_SIMPLEX, 0.9, Scalar(0, 255, 255), 2, cv::LINE_AA);
            imwrite("out.jpg",frameFace);
        }
        
    }
}
tuple<Mat, vector<vector<int>>> ImagePredictor::getFaceBox(Net net, Mat &frame, double conf_threshold)
{
    Mat frameOpenCVDNN = frame.clone();
    cout<<"clone done" <<endl;
    int frameHeight = frameOpenCVDNN.rows;
    int frameWidth = frameOpenCVDNN.cols;
    double inScaleFactor = 1.0;
    Size size = Size(300, 300);
    // std::vector<int> meanVal = {104, 117, 123};
    Scalar meanVal = Scalar(104, 117, 123);

    Mat inputBlob;
    inputBlob = cv::dnn::blobFromImage(frameOpenCVDNN, inScaleFactor, size, meanVal, true, false);
    cout<<"input Blob : " << inputBlob.dims <<endl;
    net.setInput(inputBlob, "data");
    Mat detection = net.forward("detection_out");

    Mat detectionMat(detection.size[2], detection.size[3], CV_32F, detection.ptr<float>());

    vector<vector<int>> bboxes;

    for(int i = 0; i < detectionMat.rows; i++)
    {
        float confidence = detectionMat.at<float>(i, 2);

        if(confidence > conf_threshold)
        {
            int x1 = static_cast<int>(detectionMat.at<float>(i, 3) * frameWidth);
            int y1 = static_cast<int>(detectionMat.at<float>(i, 4) * frameHeight);
            int x2 = static_cast<int>(detectionMat.at<float>(i, 5) * frameWidth);
            int y2 = static_cast<int>(detectionMat.at<float>(i, 6) * frameHeight);
            vector<int> box = {x1, y1, x2, y2};
            bboxes.push_back(box);
            rectangle(frameOpenCVDNN, Point(x1, y1), Point(x2, y2), Scalar(0, 255, 0),2, 4);
        }
    }

    return make_tuple(frameOpenCVDNN, bboxes);
}



ImagePredictor* ImagePredictor_new(
    const char * faceProto,
    const char * faceModel,
    const char * ageProto,
    const char * ageModel,
    const char * genderProto,
    const char * genderModel
    )
{
    return new ImagePredictor(
            faceProto,
            faceModel,
            ageProto,
            ageModel,
            genderProto,
            genderModel
        );
}

void ImagePredictor_predict(ImagePredictor *ip,int rows, int cols, unsigned char *buffer)
{
    ip->predict(rows, cols, buffer);
}

File: inference-engine/handler/ImagePredictor.py
# Import required modules
import time
import cv2 as cv


MODEL_MEAN_VALUES = (78.4263377603, 87.7689143744, 114.895847746)
ageList = ['(0-2)', '(4-6)', '(8-12)', '(15-20)', '(25-32)', '(38-43)', '(48-53)', '(60-100)']
genderList = ['Male', 'Female']

def getFaceBox(net, frame, conf_threshold=0.7):
    frame_opencv_dnn = frame.copy()
    frame_height = frame_opencv_dnn.shape[0]
    frame_width = frame_opencv_dnn.shape[1]
    blob = cv.dnn.blobFromImage(frame_opencv_dnn, 1.0, (300, 300), [104, 117, 123], True, False)

    net.setInput(blob)
    detections = net.forward()
    bboxes = []
    for i in range(detections.shape[2]):
        confidence = detections[0, 0, i, 2]
        if confidence > conf_threshold:
            x1 = int(detections[0, 0, i, 3] * frame_width)
            y1 = int(detections[0, 0, i, 4] * frame_height)
            x2 = int(detections[0, 0, i, 5] * frame_width)
            y2 = int(detections[0, 0, i, 6] * frame_height)
            bboxes.append([x1, y1, x2, y2])
            cv.rectangle(
                frame_opencv_dnn, (x1, y1), (x2, y2), (0, 255, 0), int(round(frame_height/150)), 8
            )

    return frame_opencv_dnn, bboxes

class ImagePredictor:
    def __init__(self):
        faceProto = "models/opencv_face_detector.pbtxt"
        faceModel = "models/opencv_face_detector_uint8.pb"

        ageProto = "models/age_deploy.prototxt"
        ageModel = "models/age_net.caffemodel"

        genderProto = "models/gender_deploy.prototxt"
        genderModel = "models/gender_net.caffemodel"

        # Load network
        self.ageNet=cv.dnn.readNet(ageModel, ageProto)
        self.genderNet=cv.dnn.readNet(genderModel, genderProto)
        self.faceNet=cv.dnn.readNet(faceModel, faceProto)       

    def read_image(self, path):
        try:
            cap = cv.VideoCapture(path)
            r, frame = cap.read()
            if r:
                return frame
                
        except Exception as e:
            print (str(e))
        
        return None

    def predict(self, frame):
        padding = 20
        # Read frame
        t = time.time()
        ret = []
        frameFace, bboxes = getFaceBox(self.faceNet, frame)
        if not bboxes:
            print("No face Detected, Checking next frame")
        else:

            for bbox in bboxes:
                # print(bbox)
                face = frame[
                    max( 0, bbox[1]-padding):min(bbox[3]+padding,
                        frame.shape[0]-1),max(0,bbox[0]-padding
                    ):min(bbox[2]+padding, frame.shape[1]-1)]

                blob = cv.dnn.blobFromImage(face, 1.0, (227, 227), MODEL_MEAN_VALUES, swapRB=False)
                self.genderNet.setInput(blob)
                genderPreds = self.genderNet.forward()
                gender = genderList[genderPreds[0].argmax()]
                print("Gender Output : {}".format(genderPreds))
                print("Gender : {}, conf = {:.3f}".format(gender, genderPreds[0].max()))

                self.ageNet.setInput(blob)
                agePreds = self.ageNet.forward()
                age = ageList[agePreds[0].argmax()]
                print("Age Output : {}".format(agePreds))
                print("Age : {}, conf = {:.3f}".format(age, agePreds[0].max()))
                

                label = "{},{}".format(gender, age)
                ret.append({"age": age, "gender": gender})
            
            cv.putText(frameFace, label, (bbox[0], bbox[1]-10), cv.FONT_HERSHEY_SIMPLEX, 0.8, (0, 255, 255), 2, cv.LINE_AA)
            cv.imwrite("age-gender.jpg", frameFace)
        print("time : {:.3f}".format(time.time() - t))

        return ret